lib/improver/utilities/indexing_operations.py

```diff
@@ -0,0 +1,157 @@
+
+       The form of the index_arrays after they have been broadcast to match the
+       sub-arrays from which values are being taken.
+
+    Args:
+        index_array (np.array of ints):
+            This array must contain integers in the range [0, n-1], where n is
+            the number of arrays in the array_set (equivalent to the length of
+            the leading dimension of array_set if it is a multi-dimensional
+            array rather than a list).
+        array_set (np.array):
+            A multi-dimensional array, where the leading dimension is in effect
+            an indexing dimension. Within this leading dimension are the
+            sub-arrays from which values are to be extracted at positions that
+            match those given in the index_array.
+    Returns:
+        result (np.array):
+            An array containing data extracted from the array_set.
+    Raises:
+        ValueError: If input arrays cannot be broadcast to a common shape.
+        IndexError: If an index exceeds the number of available sub-arrays.
+    """
+    # Used to correct array shape in output following intermediate broadcast.
+    index_array = np.array(index_array)
+    index_dims = index_array.ndim
+    trim_result = index_array.shape != array_set.shape
+
+    # Broadcast arrays to matching shape. This is included to replicate as much
+    # of the numpy choose functionality as possible.
+    try:
+        broadcast_arrays = np.broadcast_arrays(index_array, array_set)
+    except ValueError as err:
+        msg = '{}\nindex_array shape: {}\narray_set shape: {}'.format(
+            err, index_array.shape, array_set.shape)
+        raise ValueError(msg)
+
+    # numpy choose can handle the condition below, but we currently have no
+    # need to accommodate this behaviour, so we trap it and raise an error.
+    if broadcast_arrays[1].ndim > array_set.ndim:
+        msg = ('Dimensionality of array_set has increased which will prevent '
+               'indexing from working as expected.')
+        raise IndexError(msg)
+
+    index_array = broadcast_arrays[0]
+    array_set = broadcast_arrays[1]
+
+    # Use indexing to extract the requires values from the sub-arrays.
+    try:
+        result = np.array(
+            [array_set[index_array[I]][I[1:]]
+             for I in np.lib.index_tricks.ndindex(index_array.shape)]
+        ).reshape(index_array.shape)
+    except IndexError as err:
+        msg = ('{}\nindex_array contains an index that is larger than the '
+               'number of sub-arrays from which data can be drawn.\nMax '
+               'index: {}\nNumber of sub-arrays: {}'.format(
+                   err, index_array.max(), array_set.shape[0]))
+        raise IndexError(msg)
+
+    # If the index_array shape did not match the array_set shape, it is assumed
+    # that the indexing is for the sub-arrays, not the complete array_set. The
+    # leading dimension is thus cut away in the result.
```

Comment:
This links to the comment above about explaining the shape of the output array. I think this behaviour is worth flagging up.

Reply:
Done.